checker/src/org/checkerframework/checker/minlen/MinLenAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,299 @@
+                return rhsVal >= lhsVal;
+            }
+            return false;
+        }
+    }
+
+    @Override
+    protected TypeAnnotator createTypeAnnotator() {
+        return new ListTypeAnnotator(new MinLenTypeAnnotator(this), super.createTypeAnnotator());
+    }
+
+    protected class MinLenTypeAnnotator extends TypeAnnotator {
+
+        public MinLenTypeAnnotator(MinLenAnnotatedTypeFactory atf) {
+            super(atf);
+        }
+
+        @Override
+        public Void visitArray(AnnotatedArrayType type, Void aVoid) {
+            if (!type.hasAnnotation(MinLen.class)) {
+                type.replaceAnnotation(MIN_LEN_0);
+            }
+            return super.visitArray(type, aVoid);
+        }
+    }
+
+    @Override
+    public TreeAnnotator createTreeAnnotator() {
+        return new ListTreeAnnotator(
+                new MinLenTreeAnnotator(this),
+                new PropagationTreeAnnotator(this),
+                new ImplicitsTreeAnnotator(this));
+    }
+
+    protected class MinLenTreeAnnotator extends TreeAnnotator {
+
+        public MinLenTreeAnnotator(MinLenAnnotatedTypeFactory factory) {
+            super(factory);
+        }
+
+        /** When a new array is reached, record how long it is. */
+        @Override
+        public Void visitNewArray(NewArrayTree tree, AnnotatedTypeMirror type) {
+
+            AnnotatedTypeMirror valueType = valueAnnotatedTypeFactory.getAnnotatedType(tree);
+
+            if (valueType.hasAnnotation(ArrayLen.class)) {
+                AnnotationMirror anm = valueType.getAnnotation(ArrayLen.class);
+                Integer val = Collections.min(ValueAnnotatedTypeFactory.getArrayLength(anm));
+                type.replaceAnnotation(createMinLen(val));
+            }
+
+            return super.visitNewArray(tree, type);
+        }
+
+        @Override
+        public Void visitLiteral(LiteralTree tree, AnnotatedTypeMirror type) {
+
+            if (tree.getKind() == Tree.Kind.STRING_LITERAL) {
+                String lit = (String) (tree.getValue());
+                type.replaceAnnotation(createMinLen(lit.length()));
+            }
+
+            return super.visitLiteral(tree, type);
+        }
+
+        @Override
+        public Void visitIdentifier(IdentifierTree tree, AnnotatedTypeMirror type) {
+            AnnotatedTypeMirror valueType = valueAnnotatedTypeFactory.getAnnotatedType(tree);
+
+            if (valueType.hasAnnotation(ArrayLen.class)) {
+                AnnotationMirror anm = valueType.getAnnotation(ArrayLen.class);
+                Integer val = Collections.min(ValueAnnotatedTypeFactory.getArrayLength(anm));
+                type.replaceAnnotation(createMinLen(val));
+            }
+            return super.visitIdentifier(tree, type);
+        }
+    }
+
+    public ValueAnnotatedTypeFactory getValueAnnotatedTypeFactory() {
+        return valueAnnotatedTypeFactory;
+    }
+
+    protected static int getMinLenValue(AnnotationMirror annotation) {
```

Comment:
The use in MinLenStore#updateForMethodCall doesn't fit this pattern. So, I'm going to leave the existing method, and add another method that abstracts the other three uses that do fit the pattern you identified.